File: tests/r_parity/62_truncreg.R
# StatsPAI truncated regression parity (R side) -- Module 62.
#
# truncreg estimates sigma on the natural scale, matching the Stata
# truncreg /sigma row; the Python side delta-maps exp(ln_sigma).

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) dirname(normalizePath(sub("^--file=", "", .file_arg[1]))) else getwd()
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({ library(truncreg) })

MODULE <- "62_truncreg"
df <- read_csv_strict(MODULE)

# method='NR' (maxLik Newton-Raphson) converges past the default BFGS
# stopping point to the same likelihood optimum found by sp.truncreg
# and Stata truncreg (logLik gain ~5e-8 over the BFGS default).
fit <- truncreg(y ~ x1 + x2, data = df, point = 0, direction = "left",
                method = "NR", iterlim = 500)
co <- coef(fit); se <- sqrt(diag(vcov(fit)))

rows <- list(
  parity_row(MODULE, "beta_intercept", estimate = unname(co["(Intercept)"]), se = unname(se["(Intercept)"]), n = nrow(df)),
  parity_row(MODULE, "beta_x1",        estimate = unname(co["x1"]),          se = unname(se["x1"]),          n = nrow(df)),
  parity_row(MODULE, "beta_x2",        estimate = unname(co["x2"]),          se = unname(se["x2"]),          n = nrow(df)),
  parity_row(MODULE, "sigma",          estimate = unname(co["sigma"]),       se = unname(se["sigma"]),       n = nrow(df))
)
write_results(MODULE, rows,
              extra = list(package = "truncreg::truncreg",
                           truncation = "left at 0"))

File: tests/r_parity/62_truncreg.py
"""StatsPAI truncated regression parity (Python side) -- Module 62.

Left-truncated (at 0) normal regression against R truncreg::truncreg
and Stata truncreg, ll(0). sp.truncreg estimates ln_sigma; the sigma
row is reported on the natural scale with the delta-method SE so it
can be compared to the R/Stata direct-sigma parameterisation.
"""
from __future__ import annotations
import numpy as np, pandas as pd, statspai as sp
from _common import PARITY_SEED, ParityRecord, dump_csv, write_results

MODULE = "62_truncreg"


def make_data(n=1400, seed=PARITY_SEED):
    rng = np.random.default_rng(seed)
    x1 = rng.normal(0, 1, n)
    x2 = rng.uniform(0, 1, n)
    ystar = 1.0 + 0.8 * x1 - 0.5 * x2 + rng.normal(0, 1.2, n)
    keep = ystar > 0.0
    return pd.DataFrame({"y": ystar[keep], "x1": x1[keep], "x2": x2[keep]})


def main():
    df = make_data()
    dump_csv(df, MODULE)
    res = sp.truncreg(data=df, y="y", x=["x1", "x2"], ll=0.0, maxiter=2000, tol=1e-12)
    rows = []
    for nm, lab in [("_cons", "intercept"), ("x1", "x1"), ("x2", "x2")]:
        if nm in res.params.index:
            rows.append(ParityRecord(MODULE, "py", f"beta_{lab}",
                estimate=float(res.params[nm]),
                se=float(res.std_errors[nm]),
                n=int(len(df))))
    if "ln_sigma" in res.params.index:
        ln_sigma = float(res.params["ln_sigma"])
        se_ln = float(res.std_errors["ln_sigma"])
        sigma = float(np.exp(ln_sigma))
        rows.append(ParityRecord(MODULE, "py", "sigma",
            estimate=sigma, se=sigma * se_ln, n=int(len(df))))
    write_results(MODULE, "py", rows,
                  extra={"truncation": "left at 0",
                         "sigma_note": ("sp.truncreg estimates ln_sigma; sigma row is "
                                        "exp(ln_sigma) with the delta-method SE.")})


if __name__ == "__main__":
    main()
